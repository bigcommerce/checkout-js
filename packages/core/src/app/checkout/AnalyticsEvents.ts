let boltTracker: Window["BoltTrack"] = {
  recordEvent: (event: string) => console.log("--bolt bigc--: NOOP ", event),
};

const eventLog: string[] = [];

export const AnalyticsEvents = {
  init: () => {
    try {
      console.log("--bolt bigc--: intializing Bolt-BigC analytics events");
      if (window && window.BoltTrack) {
        boltTracker = window.BoltTrack;
        console.log("--bolt bigc--: successfully assigned window BoltTrack");
        // immediately send checkout load success event
        AnalyticsEvents.emitEvent("Checkout load success");
      }
    } catch (e) {
      console.log("--bolt bigc--: error during emitting event ", e);
    }
  },
  emitEvent: (eventName: string) => {
    try {
      console.log("--bolt bigc--: emitting analytics event ", eventName);
      const props: any = {
        nextState: eventName,
        prevState: "",
      };
      if (eventLog.length > 0) {
        props.prevState = eventLog[0];
      }
      boltTracker.recordEvent("CheckoutFunnelTransition", props);
      // add latest event to beginning of log array
      eventLog.unshift(eventName);
    } catch (e) {
      console.log("--bolt bigc--: error during emitting event ", e);
    }
  },
  onBeforeUnload: () => {
    AnalyticsEvents.emitEvent("Exit");
  },
};
