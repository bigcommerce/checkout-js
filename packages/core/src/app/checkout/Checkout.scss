@import '../ui/Base';

$color-secondary: #757575;
$color-link: #476BEF;
$fontFamily-montserrat: 'Montserrat';

// optimizedCheckout-headingPrimary
.header, .header-secondary, .sub-header {
    color: $color-black;
    font-family: $fontFamily-montserrat;
    font-size: fontSize("large");
    font-weight: fontWeight("bold");
}

.header-secondary {
    color: $color-secondary;
}

// optimizedCheckout-headingSecondary
.sub-header {
    font-size: fontSize("smaller");
}

// optimizedCheckout-contentPrimary
.body-regular, .body-cta, .body-medium, .body-bold {
    color: $color-black;
    font-family: $fontFamily-montserrat;
    font-size: fontSize("base");
    font-weight: fontWeight("normal");
}

.body-cta {
    color: $color-link;
}

.body-medium, .body-bold {
    font-weight: fontWeight("medium");
}

.body-bold {
    font-size: fontSize("base");
    font-weight: fontWeight("bold");
}

.sub-text, .sub-text-medium, .sub-text-bold {
    color: $color-black;
    font-family: $fontFamily-montserrat;
    font-size: fontSize("smallest");
    font-weight: fontWeight("normal");
}

.sub-text-medium {
    color: $color-secondary;
    font-weight: fontWeight("medium");
}

.sub-text-bold {
    font-weight: fontWeight("bold");
}

// overrides 
.body-bold, .body-medium {
    &.button--primary:not(:disabled){
        color: $color-white;
    }
}

.form-checkbox + .form-label,
.form-radio + .form-label {
    &.body-regular {
        color: $color-black;
        font-family: $fontFamily-montserrat;
        font-size: fontSize("base");
        font-weight: fontWeight("normal");
    }
}

.shippingOptionLabel {
    .shippingOption-price, .shippingOption-desc {
        &.body-medium {
            color: $color-black;
            font-family: $fontFamily-montserrat;
            font-size: fontSize("base");
            font-weight: fontWeight("medium");
        }
    }
}

.body-medium {
     small {
        color: $color-black;
        font-family: $fontFamily-montserrat;
        font-size: fontSize("base");
        font-weight: fontWeight("medium");
    }
}

.add-consignment-button {
    &.body-regular {
        color: $color-black;
    }
}
