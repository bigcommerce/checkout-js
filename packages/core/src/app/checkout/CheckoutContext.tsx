import { CheckoutSelectors, CheckoutService } from '@bigcommerce/checkout-sdk';
import { createContext } from 'react';

export interface CheckoutContextProps {
    checkoutService: CheckoutService;
    checkoutState: CheckoutSelectors;
}

const CheckoutContext = createContext<CheckoutContextProps | undefined>(undefined);

export default CheckoutContext;
