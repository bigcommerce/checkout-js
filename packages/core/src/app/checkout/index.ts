export { RenderCheckout, RenderCheckoutOptions } from './renderCheckout';
export { default as CheckoutContext, CheckoutContextProps } from './CheckoutContext';
export {
    default as CheckoutProvider,
    CheckoutProviderProps,
    CheckoutProviderState,
} from './CheckoutProvider';
export { default as withCheckout, WithCheckoutProps } from './withCheckout';
export { default as CheckoutSupport } from './CheckoutSupport';
export { default as NoopCheckoutSupport } from './NoopCheckoutSupport';
export { default as navigateToOrderConfirmation } from './navigateToOrderConfirmation';
