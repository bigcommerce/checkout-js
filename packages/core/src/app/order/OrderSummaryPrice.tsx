import classNames from 'classnames';
import React, { Component, ReactNode } from 'react';
import { CSSTransition } from 'react-transition-group';

import { preventDefault } from '../common/dom';
import { ShopperCurrency } from '../currency';

export interface OrderSummaryPriceProps {
    label: ReactNode;
    amount?: number | null;
    zeroLabel?: ReactNode;
    className?: string;
    testId?: string;
    currencyCode?: string;
    superscript?: string;
    actionLabel?: ReactNode;
    onActionTriggered?(): void;
}

export interface OrderSummaryPriceState {
    highlight: boolean;
    previousAmount?: number;
}

function getDisplayValue(amount?: number | null, zeroLabel?: ReactNode): ReactNode | number {
    const notYetSetSymbol = '--';

    if (typeof amount === 'undefined' || amount === null) {
        return notYetSetSymbol;
    }

    if (zeroLabel && amount === 0) {
        return zeroLabel;
    }

    return amount;
}

function isNumberValue(displayValue: number | ReactNode): displayValue is number {
    return typeof displayValue === 'number';
}

class OrderSummaryPrice extends Component<OrderSummaryPriceProps, OrderSummaryPriceState> {
    static getDerivedStateFromProps(props: OrderSummaryPriceProps, state: OrderSummaryPriceState) {
        return {
            highlight: props.amount !== state.previousAmount,
            previousAmount: props.amount,
        };
    }

    state = {
        highlight: false,
        previousAmount: 0,
    };

    render(): ReactNode {
        const {
            amount,
            actionLabel,
            onActionTriggered,
            children,
            className,
            currencyCode,
            label,
            superscript,
            testId,
            zeroLabel,
        } = this.props;

        const { highlight } = this.state;
        const displayValue = getDisplayValue(amount, zeroLabel);

        return (
            <div data-test={testId}>
                <CSSTransition
                    addEndListener={this.handleTransitionEnd}
                    classNames="changeHighlight"
                    in={highlight}
                    timeout={{}}
                >
                    <div
                        aria-live="polite"
                        className={classNames(
                            'cart-priceItem',
                            'optimizedCheckout-contentPrimary',
                            className,
                        )}
                    >
                        <span className="cart-priceItem-label">
                            <span data-test="cart-price-label">
                                {label}
                                {'  '}
                            </span>
                            {currencyCode && (
                                <span className="cart-priceItem-currencyCode">
                                    {`(${currencyCode}) `}
                                </span>
                            )}
                            {onActionTriggered && actionLabel && (
                                <span className="cart-priceItem-link">
                                    <a
                                        data-test="cart-price-callback"
                                        href="#"
                                        onClick={preventDefault(onActionTriggered)}
                                    >
                                        {actionLabel}
                                    </a>
                                </span>
                            )}
                        </span>

                        <span className="cart-priceItem-value">
                            <span data-test="cart-price-value">
                                {isNumberValue(displayValue) ? (
                                    <ShopperCurrency amount={displayValue} />
                                ) : (
                                    displayValue
                                )}
                            </span>

                            {superscript && (
                                <sup data-test="cart-price-value-superscript">{superscript}</sup>
                            )}
                        </span>

                        {children}
                    </div>
                </CSSTransition>
            </div>
        );
    }

    private handleTransitionEnd: (node: HTMLElement, done: () => void) => void = (node, done) => {
        const { previousAmount } = this.state;

        node.addEventListener('animationend', ({ target }) => {
            if (target === node) {
                this.setState({
                    highlight: false,
                    previousAmount,
                });
                done();
            }
        });
    };
}

export default OrderSummaryPrice;
