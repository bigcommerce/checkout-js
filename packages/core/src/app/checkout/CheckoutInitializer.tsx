import {
    type CheckoutInitialState,
    type EmbeddedCheckoutMessenger,
    type EmbeddedCheckoutMessengerOptions
} from '@bigcommerce/checkout-sdk';
import React from 'react';

import { useThemeContext } from '@bigcommerce/checkout/contexts';
import { type ErrorLogger } from '@bigcommerce/checkout/error-handling-utils';
import { CheckoutPageSkeleton } from '@bigcommerce/checkout/ui';

import { type EmbeddedCheckoutStylesheet } from '../embeddedCheckout';

import CheckoutPage from './CheckoutPage';
import type CheckoutSupport from './CheckoutSupport';
import { useLoadCheckout } from './hooks';

export interface CheckoutInitializerProps {
    checkoutId: string;
    containerId: string;
    embeddedStylesheet: EmbeddedCheckoutStylesheet;
    embeddedSupport: CheckoutSupport;
    errorLogger: ErrorLogger;
    initialState?: CheckoutInitialState;
    createEmbeddedMessenger(options: EmbeddedCheckoutMessengerOptions): EmbeddedCheckoutMessenger;
}

const CheckoutInitializer:React.FC<CheckoutInitializerProps>= (props) => {
    const { checkoutId, initialState } = props;
    const { isLoadingCheckout } = useLoadCheckout(checkoutId, initialState);
    const { themeV2 } = useThemeContext();

    if (isLoadingCheckout) {
        return <CheckoutPageSkeleton />;
    }

    return <CheckoutPage
            {...props}
            createEmbeddedMessenger={props.createEmbeddedMessenger}
            embeddedStylesheet={props.embeddedStylesheet}
            embeddedSupport={props.embeddedSupport}
            errorLogger={props.errorLogger}
            themeV2={themeV2}
        />;
};

export default CheckoutInitializer;
