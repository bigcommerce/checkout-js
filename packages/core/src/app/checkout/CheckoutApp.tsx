import { createCheckoutService, createEmbeddedCheckoutMessenger, createStepTracker, StepTracker } from '@bigcommerce/checkout-sdk';
import { BrowserOptions } from '@sentry/browser';
import React, { Component } from 'react';
import ReactModal from 'react-modal';

import '../../scss/App.scss';
import { createErrorLogger, ErrorBoundary, ErrorLogger } from '../common/error';
import { createEmbeddedCheckoutStylesheet, createEmbeddedCheckoutSupport } from '../embeddedCheckout';
import { getLanguageService, LocaleProvider } from '../locale';

import Checkout from './Checkout';
import CheckoutProvider from './CheckoutProvider';

export interface CheckoutAppProps {
    checkoutId: string;
    containerId: string;
    publicPath?: string;
    sentryConfig?: BrowserOptions;
}

export default class CheckoutApp extends Component<CheckoutAppProps> {
    private checkoutService = createCheckoutService({
        locale: getLanguageService().getLocale(),
        shouldWarnMutation: process.env.NODE_ENV === 'development',
    });
    private embeddedStylesheet = createEmbeddedCheckoutStylesheet();
    private embeddedSupport = createEmbeddedCheckoutSupport(getLanguageService());
    private errorLogger: ErrorLogger;

    constructor(props: Readonly<CheckoutAppProps>) {
        super(props);

        this.errorLogger = createErrorLogger(
            { sentry: props.sentryConfig },
            {
                errorTypes: ['UnrecoverableError'],
                publicPath: props.publicPath,
            }
        );
    }

    componentDidMount(): void {
        const { containerId } = this.props;

        ReactModal.setAppElement(`#${containerId}`);
    }

    render() {
        return (
            <ErrorBoundary logger={ this.errorLogger }>
                <LocaleProvider checkoutService={ this.checkoutService }>
                    <CheckoutProvider checkoutService={ this.checkoutService }>
                        <Checkout
                            { ...this.props }
                            createEmbeddedMessenger={ createEmbeddedCheckoutMessenger }
                            createStepTracker={ this.createStepTracker }
                            embeddedStylesheet={ this.embeddedStylesheet }
                            embeddedSupport={ this.embeddedSupport }
                            errorLogger={ this.errorLogger }
                        />
                    </CheckoutProvider>
                </LocaleProvider>
            </ErrorBoundary>
        );
    }

    private createStepTracker: () => StepTracker = () => {
        return createStepTracker(this.checkoutService);
    };
}
