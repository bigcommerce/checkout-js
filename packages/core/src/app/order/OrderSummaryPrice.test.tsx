import React, { ReactNode } from 'react';

import * as paymentIntegrationApi from '@bigcommerce/checkout/payment-integration-api';
import { fireEvent, render, screen } from '@bigcommerce/checkout/test-utils';

jest.mock('react-transition-group', () => ({
    CSSTransition: ({ children }: { children: React.ReactNode }) => children,
}));

jest.mock('../currency', () => ({
    ShopperCurrency: ({ amount }: {amount: number}) => <div data-test="ShopperCurrency">{amount}</div>
}));

import OrderSummaryPrice, { OrderSummaryPriceProps } from './OrderSummaryPrice';

describe('OrderSummaryPrice', () => {
    const useCheckoutMock = (isSubmittingOrder: boolean) => {
        jest.spyOn(paymentIntegrationApi, 'useCheckout').mockImplementation(
            jest.fn().mockImplementation(() => ({
                checkoutState: {
                    statuses: {
                        isSubmittingOrder: () => isSubmittingOrder,
                    }
                }
            }))
        );
    };

    const renderTestComponent = (
        props: OrderSummaryPriceProps & { children?: ReactNode }
    ) => {
        return render(<OrderSummaryPrice {...props} />);
    }

    describe('when has non-zero amount', () => {
        const amount = 10;

        beforeEach(() => {
            useCheckoutMock(false);
        });

        describe('and has only required props', () => {
            it('renders component', () => {
                const { baseElement } = renderTestComponent({
                    amount,
                    label: 'Label',
                    children: 'Foo Children',
                });

                expect(baseElement).toMatchSnapshot();
            });
        });

        describe('and has only required props', () => {
            it('renders additional elements/props', () => {
                const { baseElement } = renderTestComponent({
                    amount,
                    className: 'extra-class',
                    currencyCode: 'EUR',
                    label: 'Label',
                    superscript: 'superscript',
                    testId: 'test-id',
                });

                expect(baseElement).toMatchSnapshot();

                expect(screen.getByText('(EUR)')).toBeInTheDocument();
                expect(screen.getByTestId('cart-price-value-superscript')).toHaveTextContent('superscript');
            });
        });
    });

    describe('when has null amount', () => {
        it('renders not yet symbol as label', () => {
            renderTestComponent({
                amount: null,
                label: 'Label',
            });

            expect(screen.getByTestId('cart-price-value')).toHaveTextContent('--');
        });
    });

    describe('when has zero amount', () => {
        const amount = 0;

        describe('and no label', () => {
            it('renders formatted amount', () => {
                renderTestComponent({
                    amount,
                    className: 'label',
                    label: 'Label',
                    testId: 'test',
                });

                expect(screen.getByTestId('ShopperCurrency')).toHaveTextContent('0');
            });
        });

        describe('and zero label', () => {
            it('renders zero label', () => {
                renderTestComponent({
                    amount,
                    className: 'label',
                    label: 'Label',
                    testId: 'test',
                    zeroLabel: 'Free',
                });

                expect(screen.getByTestId('cart-price-value')).toHaveTextContent('Free');
            });
        });
    });

    describe('price action button', () => {
        const amount = 10;
        let onActionTriggered: () => void;
        let actionLabel: string;

        beforeEach(() => {
            onActionTriggered = jest.fn();
            actionLabel = 'Action Label';
        });

        it('should show price action button', () => {
            renderTestComponent({
                amount,
                className: 'label',
                label: 'Label',
                testId: 'test',
                onActionTriggered,
                actionLabel,
            });

            const actionButton = screen.getByTestId('cart-price-callback');

            expect(actionButton).toBeInTheDocument();
            expect(actionButton).toHaveTextContent(actionLabel);

            fireEvent.click(actionButton);

            expect(onActionTriggered).toHaveBeenCalled();
        });

        it('should not show price action button if no callback for this action', () => {
            renderTestComponent({
                amount,
                className: 'label',
                label: 'Label',
                testId: 'test',
                onActionTriggered: undefined,
                actionLabel,
            });

            expect(screen.queryByTestId('cart-price-callback')).not.toBeInTheDocument();
        });

        it('should not show price action button if no label for this button', () => {
            renderTestComponent({
                amount,
                className: 'label',
                label: 'Label',
                testId: 'test',
                onActionTriggered,
                actionLabel: undefined,
            });

            expect(screen.queryByTestId('cart-price-callback')).not.toBeInTheDocument();
        });

        it('should not show price action button while submitting payment', () => {
            useCheckoutMock(true);

            renderTestComponent({
                amount,
                className: 'label',
                label: 'Label',
                testId: 'test',
                onActionTriggered,
                actionLabel,
            });

            const actionButton = screen.getByTestId('cart-price-callback');

            expect(actionButton).toBeInTheDocument();
            expect(actionButton).toHaveTextContent(actionLabel);

            fireEvent.click(actionButton);

            expect(onActionTriggered).not.toHaveBeenCalled();
        });
    });
});
