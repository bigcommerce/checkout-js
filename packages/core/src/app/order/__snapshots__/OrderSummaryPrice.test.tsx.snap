// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderSummaryPrice when has non-zero amount and has only required props renders additional elements/props 1`] = `
<body>
  <div>
    <div
      data-test="test-id"
    >
      <div
        aria-live="polite"
        class="cart-priceItem optimizedCheckout-contentPrimary extra-class"
      >
        <span
          class="cart-priceItem-label"
        >
          <span
            data-test="cart-price-label"
          >
            Label
              
          </span>
          <span
            class="cart-priceItem-currencyCode"
          >
            (EUR) 
          </span>
        </span>
        <span
          class="cart-priceItem-value"
        >
          <span
            data-test="cart-price-value"
          >
            <div
              data-test="ShopperCurrency"
            >
              10
            </div>
          </span>
          <sup
            data-test="cart-price-value-superscript"
          >
            superscript
          </sup>
        </span>
      </div>
    </div>
  </div>
</body>
`;

exports[`OrderSummaryPrice when has non-zero amount and has only required props renders component 1`] = `
<body>
  <div>
    <div>
      <div
        aria-live="polite"
        class="cart-priceItem optimizedCheckout-contentPrimary"
      >
        <span
          class="cart-priceItem-label"
        >
          <span
            data-test="cart-price-label"
          >
            Label
              
          </span>
        </span>
        <span
          class="cart-priceItem-value"
        >
          <span
            data-test="cart-price-value"
          >
            <div
              data-test="ShopperCurrency"
            >
              10
            </div>
          </span>
        </span>
        Foo Children
      </div>
    </div>
  </div>
</body>
`;
